## Lý do chọn thuật toán Gaussian Naive Bayes

**Gaussian Naive Bayes (GNB)** được lựa chọn vì các lý do sau:

- **Phù hợp với dữ liệu dạng liên tục:**  
  Bộ dữ liệu *Pima Indians Diabetes Database* có các thuộc tính như `Glucose`, `BloodPressure`, `BMI`,… là các giá trị **số thực (continuous)**.  
  Thuật toán GNB giả định rằng dữ liệu trong mỗi lớp tuân theo **phân phối chuẩn (Gaussian)**, nên rất phù hợp cho loại dữ liệu này.

- **Đơn giản, dễ huấn luyện và giải thích:**  
  GaussianNB dựa trên **định lý Bayes** và giả định **độc lập giữa các đặc trưng**, giúp việc tính toán xác suất nhanh, hiệu quả và dễ hiểu — đặc biệt quan trọng trong bài toán y học cần mô hình minh bạch.

- **Hiệu quả với dữ liệu nhỏ và vừa:**  
  Bộ dữ liệu có **768 mẫu** và **8 đặc trưng**, do đó GNB huấn luyện nhanh, không cần tinh chỉnh nhiều siêu tham số.

- **Phù hợp bài toán phân loại nhị phân:**  
  Biến mục tiêu `Outcome` gồm:
  - `0`: Không bị tiểu đường  
  - `1`: Bị tiểu đường  
  ⇒ Đây là **bài toán phân loại nhị phân (binary classification)** — rất thích hợp cho GNB.

---

## 📊 Chia tập train/test

Sau khi xử lý giá trị thiếu, dữ liệu được chia thành hai phiên bản:

| Dataset | Cách xử lý giá trị thiếu |
|----------|--------------------------|
| **Dataset 1** | Thay NaN bằng **Mean** |
| **Dataset 2** | Thay NaN bằng **Median** |

Mỗi tập dữ liệu được chia thành:
- **80% train set**: dùng để huấn luyện mô hình  
- **20% test set**: dùng để đánh giá mô hình  
- Dùng `random_state=42` để tái lập kết quả.

```python
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)


In [6]:
# ===========================
# 03b_GaussianNB_Training.ipynb
# ===========================

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, precision_score, recall_score, f1_score

# ===========================
# ĐỌC DỮ LIỆU
# ===========================
df1 = pd.read_csv("../../datasets/processed/diabetes_dataset1_mean.csv")
df2 = pd.read_csv("../../datasets/processed/diabetes_dataset2_median.csv")


def evaluate_model(y_true, y_pred, model_name, dataset_name):
    print(f"\n{'='*60}")
    print(f"{model_name} - {dataset_name}")
    print(f"{'='*60}")

    acc = accuracy_score(y_true, y_pred)
    prec = precision_score(y_true, y_pred)
    rec = recall_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred)

    print(f"- Accuracy:  {acc:.4f}")
    print(f"- Precision: {prec:.4f}")
    print(f"- Recall:    {rec:.4f}")
    print(f"- F1-Score:  {f1:.4f}")
    print("\nConfusion Matrix:")
    print(confusion_matrix(y_true, y_pred))
    print("\nClassification Report:")
    print(classification_report(y_true, y_pred, target_names=['Không bị', 'Bị']))

    # return acc, prec, rec, f1

# ===========================
# TRAIN & TEST GaussianNB
# ===========================
print("\n--- Gaussian NB với Dataset 1 (Mean) ---")
X1 = df1.drop('Outcome', axis=1)
y1 = df1['Outcome']
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.2, random_state=42)
nb1 = GaussianNB()
nb1.fit(X1_train, y1_train)
y1_pred = nb1.predict(X1_test)
evaluate_model(y1_test, y1_pred, "Gaussian NB", "Dataset 1 (Mean)")

print("\n--- Gaussian NB với Dataset 2 (Median) ---")
X2 = df2.drop('Outcome', axis=1)
y2 = df2['Outcome']
X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.2, random_state=42)
nb2 = GaussianNB()
nb2.fit(X2_train, y2_train)
y2_pred = nb2.predict(X2_test)
evaluate_model(y2_test, y2_pred, "Gaussian NB", "Dataset 2 (Median)")



--- Gaussian NB với Dataset 1 (Mean) ---

Gaussian NB - Dataset 1 (Mean)
- Accuracy:  0.7468
- Precision: 0.6379
- Recall:    0.6727
- F1-Score:  0.6549

Confusion Matrix:
[[78 21]
 [18 37]]

Classification Report:
              precision    recall  f1-score   support

    Không bị       0.81      0.79      0.80        99
          Bị       0.64      0.67      0.65        55

    accuracy                           0.75       154
   macro avg       0.73      0.73      0.73       154
weighted avg       0.75      0.75      0.75       154


--- Gaussian NB với Dataset 2 (Median) ---

Gaussian NB - Dataset 2 (Median)
- Accuracy:  0.7532
- Precision: 0.6491
- Recall:    0.6727
- F1-Score:  0.6607

Confusion Matrix:
[[79 20]
 [18 37]]

Classification Report:
              precision    recall  f1-score   support

    Không bị       0.81      0.80      0.81        99
          Bị       0.65      0.67      0.66        55

    accuracy                           0.75       154
   macro avg       

In [7]:
import joblib
joblib.dump(nb1, "../../models/diabetes_nb_model.pkl")
print("✅ Mô hình đã được lưu thành công!")

✅ Mô hình đã được lưu thành công!


## 🔹 ĐÁNH GIÁ VÀ NHẬN XÉT KẾT QUẢ MÔ HÌNH GAUSSIAN NAIVE BAYES (GNB)

### 🧮 1. Kết quả huấn luyện và kiểm thử

#### **Dataset 1 (Thay giá trị thiếu bằng Mean)**
- **Accuracy:** 0.7468  
- **Precision:** 0.6379  
- **Recall:** 0.6727  
- **F1-Score:** 0.6549  

**Ma trận nhầm lẫn:**
|            | Dự đoán Không bị | Dự đoán Bị |
|-------------|------------------|-------------|
| **Thực tế Không bị** | 78 | 21 |
| **Thực tế Bị**        | 18 | 37 |

---

#### **Dataset 2 (Thay giá trị thiếu bằng Median)**
- **Accuracy:** 0.7532  
- **Precision:** 0.6491  
- **Recall:** 0.6727  
- **F1-Score:** 0.6607  

**Ma trận nhầm lẫn:**
|            | Dự đoán Không bị | Dự đoán Bị |
|-------------|------------------|-------------|
| **Thực tế Không bị** | 79 | 20 |
| **Thực tế Bị**        | 18 | 37 |

---

### 📊 2. Phân tích kết quả mô hình

Mô hình **Gaussian Naive Bayes (GNB)** là một thuật toán xác suất dựa trên **Định lý Bayes** và giả định rằng **các thuộc tính độc lập với nhau** trong cùng một lớp.  
Điều này giúp GNB có ưu điểm lớn về **tốc độ huấn luyện** và **tính đơn giản**, rất phù hợp cho những bộ dữ liệu có kích thước vừa phải như Pima Indians Diabetes Dataset.

Kết quả cho thấy mô hình đạt:
- Độ chính xác (**Accuracy**) trung bình khoảng **74–75%**,  
- Các chỉ số **Precision**, **Recall** và **F1-score** đều nằm quanh mức **0.64–0.67**, phản ánh rằng mô hình hoạt động **khá cân bằng** giữa hai lớp "Bị tiểu đường" và "Không bị tiểu đường".

---

### 🔍 3. Nhận xét chi tiết

1. **Về khả năng dự đoán:**
   - Mô hình dự đoán **chính xác khá tốt các trường hợp “Không bị”** (do lớp này chiếm tỷ lệ cao hơn trong dữ liệu).  
   - Tuy nhiên, một số **bệnh nhân thực tế “bị tiểu đường”** vẫn bị dự đoán sai thành “không bị”.  
   - Điều này cho thấy **mô hình vẫn còn thiên lệch nhẹ về phía lớp đa số**.

2. **Về hiệu quả mô hình:**
   - Với dữ liệu mean và median, độ chính xác không chênh lệch nhiều, chứng tỏ mô hình **không quá nhạy cảm với cách xử lý giá trị thiếu**.  
   - Sự ổn định này phản ánh rằng GNB có khả năng **tổng quát hóa tốt** trên dữ liệu Pima.

3. **Về bản chất của GNB:**
   - GNB giả định **các đặc trưng độc lập** – điều này **không hoàn toàn đúng trong thực tế y học**, vì các yếu tố như Glucose, BMI, Insulin, BloodPressure có mối liên hệ sinh học với nhau.  
   - Tuy vậy, mô hình vẫn cho kết quả khả quan, chứng minh rằng **giả định độc lập vẫn đủ mạnh để mang lại hiệu quả phân loại tốt trong thực tế.**

4. **Về độ phức tạp:**
   - GNB có **tốc độ huấn luyện nhanh**, không yêu cầu tinh chỉnh nhiều siêu tham số, phù hợp để **dùng làm baseline** (mô hình tham chiếu ban đầu) cho việc so sánh với các thuật toán khác như Decision Tree (ID3) hay KNN.

---

### ⚙️ 4. Tổng kết

- Gaussian Naive Bayes thể hiện **hiệu suất ổn định, độ chính xác khoảng 75%, F1-score ~0.66**.  
- Mô hình **dễ triển khai, tính toán nhanh, phù hợp dữ liệu vừa và nhỏ**.  
- Tuy có hạn chế khi các thuộc tính **không độc lập** hoàn toàn, nhưng GNB vẫn là một **mô hình nền tảng đáng tin cậy** để nhận diện nguy cơ tiểu đường trong bộ dữ liệu Pima.  


---

### 🧠 Kết luận
> Mô hình **Gaussian Naive Bayes** là lựa chọn hợp lý khi cần một giải pháp **nhanh, nhẹ, và có độ chính xác khá cao (~75%)**.  
> Dù còn một số sai lệch với lớp “bị tiểu đường”, nhưng kết quả này chứng minh rằng GNB có thể **hiệu quả trong các bài toán y học cơ bản** và là nền tảng tốt để mở rộng sang các mô hình phức tạp hơn.
